lib/ansible/module_utils/aws/elbv2.py

```diff
@@ -0,0 +1,757 @@
+
+    def compare_subnets(self):
+        """
+        Compare user subnets with current ELB subnets
+
+        :return: bool True if they match otherwise False
+        """
+
+        subnet_id_list = []
+        subnets = []
+
+        # Check if we're dealing with subnets or subnet_mappings
+        if self.subnets is not None:
+            # We need to first get the subnet ID from the list
+            subnets = self.subnets
+
+        if self.subnet_mappings is not None:
+            # Make a list from the subnet_mappings dict
+            subnets_from_mappings = []
+            for subnet_mapping in self.subnet_mappings:
+                subnets.append(subnet_mapping['SubnetId'])
+
+        for subnet in self.elb['AvailabilityZones']:
+            subnet_id_list.append(subnet['SubnetId'])
+
+        if set(subnet_id_list) != set(subnets):
+            return False
+        else:
+            return True
+
+    def modify_subnets(self):
+        """
+        Modify elb subnets to match module parameters
+        :return:
+        """
+
+        try:
+            AWSRetry.jittered_backoff()(
+                self.connection.set_subnets
+            )(LoadBalancerArn=self.elb['LoadBalancerArn'], Subnets=self.subnets)
+        except (BotoCoreError, ClientError) as e:
+            self.module.fail_json_aws(e)
+
+        self.changed = True
+
+    def update(self):
+        """
+        Update the elb from AWS
+        :return:
+        """
+
+        self.elb = get_elb(self.connection, self.module, self.module.params.get("name"))
+        self.elb['tags'] = self.get_elb_tags()
+
+
+class ApplicationLoadBalancer(ElasticLoadBalancerV2):
+
+    def __init__(self, connection, connection_ec2, module):
+        """
+
+        :param connection: boto3 connection
+        :param module: Ansible module
+        """
+        super(ApplicationLoadBalancer, self).__init__(connection, module)
+
+        self.connection_ec2 = connection_ec2
+
+        # Ansible module parameters specific to ALBs
+        self.type = 'application'
+        if module.params.get('security_groups') is not None:
+            try:
+                self.security_groups = AWSRetry.jittered_backoff()(
+                    get_ec2_security_group_ids_from_names
+                )(module.params.get('security_groups'), self.connection_ec2, boto3=True)
+            except ValueError as e:
+                self.module.fail_json(msg=str(e), exception=traceback.format_exc())
+            except (BotoCoreError, ClientError) as e:
+                self.module.fail_json_aws(e)
+        else:
+            self.security_groups = module.params.get('security_groups')
+        self.access_logs_enabled = module.params.get("access_logs_enabled")
+        self.access_logs_s3_bucket = module.params.get("access_logs_s3_bucket")
+        self.access_logs_s3_prefix = module.params.get("access_logs_s3_prefix")
+        self.idle_timeout = module.params.get("idle_timeout")
+
+    def create_elb(self):
```

Comment:
This looks fairly similar for both kinds of load balancer, other than the security_groups - could this be lifted into the parent class?

Reply:
Yes, it probably could, but for future-proofing i think it's best to leave here.  Who knows how the two types may diverge in the future or even if there will be a third or fourth type that has very different options.